hpx/lcos/dist_object.hpp

```diff
@@ -0,0 +1,566 @@
+        /// string across the localities
+        /// \param data The data of the type T of the dist_object
+        dist_object(std::string base, data_type data)
+          : base_type(create_server(data))
+          , base_(base)
+        {
+            HPX_ASSERT(C == construction_type::All_to_All ||
+                C == construction_type::Meta_Object);
+            if (C == construction_type::Meta_Object)
+            {
+                meta_object mo(base);
+                basename_list = mo.registration(this->get_id());
+                basename_registration_helper(base);
+            }
+            else
+            {
+                basename_registration_helper(base);
+            }
+        }
+
+        /// \cond NOINTERNAL
+        dist_object(hpx::future<hpx::id_type>&& id)
+          : base_type(std::move(id))
+        {
+        }
+
+        dist_object(hpx::id_type&& id)
+          : base_type(std::move(id))
+        {
+        }
+        /// \endcond
+
+        /// Access the calling locality's value instance for this dist_object
+        data_type const operator*() const
+        {
+            HPX_ASSERT(this->get_id());
+            ensure_ptr();
+            return **ptr;
+        }
+
+        /// Access the calling locality's value instance for this dist_object
+        data_type operator*()
+        {
+            HPX_ASSERT(this->get_id());
+            ensure_ptr();
+            return **ptr;
+        }
+
+        /// Access the calling locality's value instance for this dist_object
+        T const* operator->() const
+        {
+            HPX_ASSERT(this->get_id());
+            ensure_ptr();
+            return &**ptr;
+        }
+
+        /// Access the calling locality's value instance for this dist_object
+        T* operator->()
+        {
+            HPX_ASSERT(this->get_id());
+            ensure_ptr();
+            return &**ptr;
+        }
+
+        /// Asynchronously returns a future of a copy of the instance of this
+        /// dist_object associated with the given locality index. The locality
+        /// index must be a valid locality ID with this dist_object.
+        hpx::future<T> fetch(int idx)
+        {
+            HPX_ASSERT(this->get_id());
+            hpx::id_type lookup = get_basename_helper(idx);
+            typedef typename server::dist_object_part<T&>::fetch_ref_action
+                action_type;
+            return hpx::async<action_type>(lookup);
+        }
+        /// \cond NOINTERNAL
+    private:
+        mutable std::shared_ptr<server::dist_object_part<T&>> ptr;
+        std::string base_;
+        void ensure_ptr() const
+        {
+            if (!ptr)
+            {
+                ptr = hpx::get_ptr<server::dist_object_part<T&>>(
+                    hpx::launch::sync, this->get_id());
+            }
+        }
+
+    private:
+        std::vector<hpx::id_type> basename_list;
+        hpx::id_type get_basename_helper(int idx)
+        {
```

Comment:
This needs to be protected with a mutex or documented as being unsafe.

Reply:
I might miss this suggestion at this point. Do you mean that I need to place a `mutex` on `std::vector<hpx::id_type> basename_list;`. Or how I should write the code to fix it?

Reply:
Yes, you need to protect the member to be modified and read concurrently.

Reply:
We should also turn this into a lazy operation, I think.